lib/autoupdate/rollout/strategy_haltonerror.go

```diff
@@ -0,0 +1,157 @@
+	previousGroupsAreDone := true
+
+	for i, group := range groups {
+		switch group.State {
+		case autoupdate.AutoUpdateAgentGroupState_AUTO_UPDATE_AGENT_GROUP_STATE_UNSTARTED:
+			var previousGroup *autoupdate.AutoUpdateAgentRolloutStatusGroup
+			if i != 0 {
+				previousGroup = groups[i-1]
+			}
+			canStart, err := canStartHaltOnError(group, previousGroup, now)
+			if err != nil {
+				// In halt-on-error rollouts, groups are dependent.
+				// Failing to transition a group should prevent other groups from transitioning.
+				setGroupState(group, group.State, updateReasonReconcilerError, now)
+				return err
+			}
+			switch {
+			case previousGroupsAreDone && canStart:
+				// We can start
+				setGroupState(group, autoupdate.AutoUpdateAgentGroupState_AUTO_UPDATE_AGENT_GROUP_STATE_ACTIVE, updateReasonCanStart, now)
+			case previousGroupsAreDone:
+				// All previous groups are OK, but time-related criteria are not OK
+				setGroupState(group, group.State, updateReasonCannotStart, now)
+			default:
+				// At least one previous group is not DONE
+				setGroupState(group, group.State, updateReasonPreviousGroupsNotDone, now)
+			}
+			previousGroupsAreDone = false
+		case autoupdate.AutoUpdateAgentGroupState_AUTO_UPDATE_AGENT_GROUP_STATE_ROLLEDBACK:
+			// The group has been manually rolled back. We don't touch anything and
+			// don't process the next groups.
+			previousGroupsAreDone = false
+		case autoupdate.AutoUpdateAgentGroupState_AUTO_UPDATE_AGENT_GROUP_STATE_DONE:
+			// The group has already been updated, we can look at the next group
+		case autoupdate.AutoUpdateAgentGroupState_AUTO_UPDATE_AGENT_GROUP_STATE_ACTIVE:
+			// The group is currently being updated. We check if we can transition it to the done state
+			done, reason := isDoneHaltOnError(group, now)
+
+			if done {
+				// We transition to the done state. We continue processing the groups as we might be able to start the next one.
+				setGroupState(group, autoupdate.AutoUpdateAgentGroupState_AUTO_UPDATE_AGENT_GROUP_STATE_DONE, reason, now)
+			} else {
+				setGroupState(group, autoupdate.AutoUpdateAgentGroupState_AUTO_UPDATE_AGENT_GROUP_STATE_ACTIVE, reason, now)
+			}
+			previousGroupsAreDone = false
+
+		default:
+			return trace.BadParameter("unknown autoupdate group state: %v", group.State)
+		}
+	}
+	return nil
+}
+
+func canStartHaltOnError(group, previousGroup *autoupdate.AutoUpdateAgentRolloutStatusGroup, now time.Time) (bool, error) {
+	// check wait days
+	if group.ConfigWaitDays != 0 {
+		if previousGroup == nil {
+			return false, trace.BadParameter("The first group cannot have non-zero wait days")
+		}
+
+		previousStart := previousGroup.StartTime.AsTime()
+		if previousStart.IsZero() || previousStart.Unix() == 0 {
+			return false, trace.BadParameter("The previous group doesn't have a start time, cannot check the 'wait_days' criteria")
+		}
+
+		// Take the day of the previous group start, add 'wait_days' and truncate to midnight.
```

Comment:
Does midnight UTC make sense here, or should we assume `wait_days` after the window itself?

Reply:
I think the RFD says midnight UTC. As a user, if I have:
- group 1 updating at 2pm
- group 2 updating at 1pm, with wait_days = 1

I would expect group 1 to update on Monday, and group 2 to update on Tuesday. If we compare to the start the of the maintenance, this would cause group 2 to update on Wednesday.

Reply:
But you have a point regarding UTC vs non-UTC. The logic would make more sense if midnight was not UTC. I'll think about it and might switch back to a strict 24h rule.